Ajout de propriétés et de dépendances

av1m opened the pull request:
Bonjour, 

Cette PR a pour objectif : 
- L'ajout des propriétés `maven.compiler.source` et `maven.compiler.target` afin de pouvoir exécuter directement un projet.
- La mise à jour des dépendances﻿﻿ `guava` et `junit`

av1m committed:
add maven compiler source & target + update dependency version

oliviercailloux commented on src/main/resources/archetype-resources/pom.xml:
```diff
@@ -10,6 +10,8 @@
 	<properties>
 		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
 		<maven.compiler.release>11</maven.compiler.release>
+		<maven.compiler.source>11</maven.compiler.source>
+		<maven.compiler.target>11</maven.compiler.target>
```
This does not seem useful to me: release should imply source and target. Please tell me if you know otherwise and I will check in the docs.
(I know that some versions of Eclipse do not understand "release" and require "source" and "target", but I believe this has been corrected in most recent versions.)

av1m replied:
I encountered this problem on CLI.

If target and source are not specified, they default to 1.5.
We can see this by activating debug mode with `mvn -X compile`.

```shell script
 -sourcepath /home/avi/github/dbms/src/main/java: -g -nowarn -target 1.5 -source 1.5 -encoding UTF-8
```

Only, this implies to a compilation error by maven because 1.5 is no longer supported : 

<pre>[<font color="#12488B"><b>INFO</b></font>] -------------------------------------------------------------
[<font color="#C01C28"><b>ERROR</b></font>] COMPILATION ERROR : 
[<font color="#12488B"><b>INFO</b></font>] -------------------------------------------------------------
[<font color="#C01C28"><b>ERROR</b></font>] Source option 5 is no longer supported. Use 7 or later.
[<font color="#C01C28"><b>ERROR</b></font>] Target option 5 is no longer supported. Use 7 or later.
[<font color="#12488B"><b>INFO</b></font>] 2 errors 
[<font color="#12488B"><b>INFO</b></font>] -------------------------------------------------------------
[<font color="#12488B"><b>INFO</b></font>] <b>------------------------------------------------------------------------</b>
[<font color="#12488B"><b>INFO</b></font>] <font color="#C01C28"><b>BUILD FAILURE</b></font>
[<font color="#12488B"><b>INFO</b></font>] <b>------------------------------------------------------------------------</b>
[<font color="#12488B"><b>INFO</b></font>] Total time:  1.232 s
[<font color="#12488B"><b>INFO</b></font>] Finished at: 2020-12-17T17:55:56+01:00
[<font color="#12488B"><b>INFO</b></font>] <b>------------------------------------------------------------------------</b>
[<font color="#C01C28"><b>ERROR</b></font>] Failed to execute goal <font color="#26A269">org.apache.maven.plugins:maven-compiler-plugin:3.1:compile</font> <b>(default-compile)</b> on project <font color="#2AA1B3">dbms</font>: <font color="#C01C28"><b>Compilation failure</b></font>: Compilation failure: 
[<font color="#C01C28"><b>ERROR</b></font>] Source option 5 is no longer supported. Use 7 or later.
[<font color="#C01C28"><b>ERROR</b></font>] Target option 5 is no longer supported. Use 7 or later.
[<font color="#C01C28"><b>ERROR</b></font>] -&gt; <b>[Help 1]</b>
<font color="#C01C28"><b>org.apache.maven.lifecycle.LifecycleExecutionException</b></font>: <font color="#C01C28"><b>Failed to execute goal </b></font><font color="#26A269"><b>org.apache.maven.plugins:maven-compiler-plugin:3.1:compile</b></font> <b>(default-compile)</b> on project <font color="#2AA1B3">dbms</font>: <font color="#C01C28"><b>Compilation failure</b></font></pre>

I tried by indicating only the source or the target, but I still have an error which is thrown.

By doing a little research I came across [this question](https://stackoverflow.com/q/49028810)

oliviercailloux replied:
May I suggest to read [this](https://github.com/oliviercailloux/java-course/blob/master/Best%20practices/Maven.adoc), and [this answer](https://stackoverflow.com/a/59185799/) from the SO post you linked to? Can you perhaps experiment a bit with this information?

av1m replied:
I'm not sure I understood correctly: when you talk about experimenting, are you talking about this pull request?

Specifying the source and target properties can be done from the compiler plugin (by specifying only the release) :

```xml
 <plugin>
  <artifactId>maven-compiler-plugin</artifactId>
  <version>3.8.0</version>
  <configuration>
    <release>11</release>
  </configuration>
</plugin>
```

Or, an equivalent method is to put in the properties (by specifying source, target, release) :

```xml
<properties>
  <maven.compiler.release>11</maven.compiler.release>
  <maven.compiler.source>11</maven.compiler.source>
  <maven.compiler.target>11</maven.compiler.target>
</properties>
```

You [mentioned to me](https://github.com/oliviercailloux/Java-Archetype/pull/1#discussion_r545006045) (and [you wrote](https://github.com/oliviercailloux/Java-Archetype/pull/1#discussion_r545006045) in your best practices) that the release property includes the source and target property.

Only, this seems to be the case only when it comes to the plugin:

<table>
  <caption>To summarize:</caption>
  <tr>
    <td></td>
    <th scope="col">Plugin</th>
    <th scope="col">Properties</th>
  </tr>
  <tr>
    <th scope="row">release (only)</th>
    <td>Work</td>
    <td>Don't work</td>
  </tr>
  <tr>
    <th scope="row">release, source, target</th>
    <td>Work</td>
    <td>Work</td>
  </tr>
</table>

You mention in your best practices that it is better to put in the properties than to use the explicit configuration of the plugin `maven-compiler-plugin` In this case, it must be mentioned 3 lines (release, source, target) instead of just one in the plugin (release)

Plus, regarding maintaining the release attribute, [this answer](https://stackoverflow.com/a/43103038) convinced me.

oliviercailloux reviewed (changes requested):
Thank you for your PR!

oliviercailloux commented:
Thanks for the reply. I meant to check whether the information provided on SO is correct by testing it, or referring to the official docs.

I suggest to keep the release property as a property, not as a plugin configuration, and include the plugin configuration only for specifying that we want a recent version of the compiler plugin. The most recent one to date is 3.8.1, so I suggest to use this one. 

I suggest to add as a comment, just above this plugin configuration: “We require a recent version of this plugin because the maven.compiler.release property is understood only since 3.6 (see https://maven.apache.org/plugins/maven-compiler-plugin/compile-mojo.html#release), whereas the version 3.1 is currently provided by default (see https://maven.apache.org/ref/3.6.3/maven-core/default-bindings.html#Plugin_bindings_for_jar_packaging or https://maven.apache.org/ref/3-LATEST/maven-core/default-bindings.html#Plugin_bindings_for_jar_packaging).”

av1m committed:
Add maven-compiler-plugin for use release properties

We require a recent version of this plugin because the maven.compiler.release property is understood only since 3.6 (see maven.apache.org/plugins/maven-compiler-plugin/compile-mojo.html#release), whereas the version 3.1 is currently provided by default (see maven.apache.org/ref/3.6.3/maven-core/default-bindings.html#Plugin_bindings_for_jar_packaging or https://maven.apache.org/ref/3-LATEST/maven-core/default-bindings.html#Plugin_bindings_for_jar_packaging).

av1m committed:
update maven-compiler-plugin

oliviercailloux commented:
Thank you once again for this useful addition.
